### ISSUE TITLE
Predictions from Discrete Choice Models

### ISSUE THREAD

**Issue by garrettvandekamp** [vincentarelbundock/marginaleffects#1632]
When I teach my students logit and probit, the overall statistic of model fit I teach them is the Percentage of Cases Correctly Predicted (PCP) and the expected Percentage of Cases Correctly Predicted (ePCP), as discussed by Herron (1999). These model fit statistics are calculated using the predicted probabilities of every observation in the dataset. They are easy to interpret and similar in spirit to R-squared. 

In principle, PCP and ePCP can be calculated using the predictions() function in R along with some post-estimation code. But doing so would be tedious and prone to user error. I also recommend students use such model fit statistics whenever estimating ordinal and nominal regression models, but of course calculating such statistics would be even more prone to user error. There not actually much in terms of software to estimate these type of statistics. Dave Armstrong released code for the glm(), polr(), and multinom() functions, which is nice. But the code does not work for many other types of models, like logitsf() and mlogit(). So there’s a notable contribution that marginaleffects could make in terms of model interpretation.

I thought it would be nice if predictions() added a few new types to make these sort of calculations easy:

First, I think type = “probs” should be available for binary models like it is for ordinal and nominal models. Simply give the probability y=1 and y=0 as an estimate as you would for any other type of discrete choice model. I know you can get such values at present from type = “response”, but it makes it harder to teach when the type changes from “response” to “probs” for ordinal and nominal regression.

Second, I think type = “probs_ob” should be available for all discrete choice models. This would be the predicted probability of the observed value of the dependent variable for each row in your dataset. So regardless of how many categories are possible for a given dependent variable, probs_ob would calculate a single estimate. Incidentally, the mean of these estimates would actually be the ePCP that Herron (1999) describes.

Third, I think type = “response” should be reprogrammed for binary models and added for ordinal and nominal models. Instead of calculating the probability that y=1 – or any other value of the dataset – it would simply report the value of the dependent variable with the greatest probability for each row in the dataset. This would be a true predicted value for discrete choice models. And you could use it to calculate PCP pretty easily.

I’d be happy to write up some code to show what I mean by each of these things for binary choice models! That way, it might be easier to implement these suggestions for the more complicated discrete choice models. Also, here’s the citation to the paper I was talking about:

Herron, Michael C. 1999. "Postestimation uncertainty in limited dependent variable models." Political Analysis 8(1): 83-98.

**Comment by vincentarelbundock** [comment]
There are several feature requests in there, so I’ll treat them one by one.

# Goodness-of-fit statistics for discrete choice models

I agree that these statistics are useful, but I feel they are pretty squarely out of scope of `marginaleffects`.

Have you looked at the `performance` package? I’m not sure it’ll cover all your needs, but they seem to have some of that functionality, and the team is usually *very* responsive to feature requests.

https://easystats.github.io/performance/reference/performance_pcp.html?q=pcp#null

# Consistent type names across models

I was once part of an effort to create a `predictions()`-style function with consistent `type` argument across a wide range of models. It was an absolute nightmare. First, terminology is suuuper hard and inconisistent; there were endless debates about naming conventions. Second, it requires a ton of work because the dev needs to understand (pretty deeply) every single one of the supported models.

I understand that it would be nice to have consistent names for the `type` argument, but I don’t think it’s feasible for a package that aims to support the work of dozens of different statisticians and developers who each have their own preferred (sometimes conflicting) terminology.

For that reason, I chose to defer to the terminlogy used in the fitting packages.

In addition to simplicity, this has the advantage of making things more consistent for users who are already familiar with the terminology in that package. For example, if you know `multinom` well, you know you can use `type="probs"`.

Finally, a nice (documented) feature is that you can call `predictions(mod,type="trash")` and the error message will tell you exactly what the valid values are.

# Rename `type="response"` to `type="class"` for binary models

I don’t think this is a good idea because it would create an inconsistency between `predict(mod, type="response")` and `predictions(mod, type="response")`. I designed the latter to be as consistent as possible with base R, and inconsistency could be confusing to users.

# Add `type="class"` for more models

So far, I have resisted adding new types in `marginaleffects`, because this typically requires writing custom `predict()` functions, and because those functions can often be very complicated (ex: dealing with offsets, detecting link functions to only apply in appropriate cases, and various other package and model-specific features). I would much prefer for prediction types to be implemented upstream, because that could be a black hole of work.

Note that some—but not all—model-fitting packages already support `type="class"` for this purpose.

And honestly, I think it’s pretty easy to post-process to get the class predictions. Here are some `tidyverse` one-liners:

``` r
library(marginaleffects)
library(tidyverse)
mod <- MASS::polr(factor(gear) ~ mpg + wt, data = mtcars)
predictions(mod) |> filter(estimate == max(estimate), .by = rowid)
```
     Group Estimate Std. Error    z Pr(>|z|)    S 2.5 % 97.5 %
         3    0.542      0.124 4.38  < 0.001 16.3 0.299  0.784
         3    0.609      0.116 5.26  < 0.001 22.7 0.382  0.837
         3    0.605      0.116 5.21  < 0.001 22.4 0.378  0.833
         3    0.567      0.163 3.49  < 0.001 11.0 0.248  0.886
         3    0.612      0.182 3.35  < 0.001 10.3 0.254  0.969
    --- 22 rows omitted. See ?print.marginaleffects ---
         4    0.577      0.119 4.84  < 0.001 19.5 0.343  0.811
         4    0.568      0.119 4.76  < 0.001 19.0 0.334  0.801
         5    0.593      0.188 3.15  0.00164  9.3 0.224  0.961
         5    0.473      0.169 2.80  0.00504  7.6 0.142  0.803
         5    0.656      0.189 3.47  < 0.001 10.9 0.286  1.026
    Type: probs

``` r
mod <- glm(vs ~ mpg + wt, data = mtcars, family = binomial)
predictions(mod) |> mutate(estimate = as.numeric(estimate > 0.5))
```
     Estimate Pr(>|z|)   S   2.5 % 97.5 %
            0   0.9898 0.0 0.20389  0.793
            1   0.8010 0.3 0.27942  0.773
            1   0.3546 1.5 0.29971  0.914
            1   0.4141 1.3 0.31724  0.866
            0   0.1585 2.7 0.14685  0.571
    --- 22 rows omitted. See ?print.marginaleffects ---
            1   0.0179 5.8 0.67628  1.000
            0   0.0373 4.7 0.00915  0.464
            0   0.4009 1.3 0.11828  0.691
            0   0.0128 6.3 0.00951  0.365
            1   0.6183 0.7 0.29678  0.810

**Comment by vincentarelbundock** [comment]
> Moving forward, I might see if other packages build additional predict() functionality for popular functions. Were these to exist, would you be open to adding them to the supported models in marginaleffects? Or would it need to be first party support? I'd like to know where you head is at in terms of development.

Yeah, I think we could build something on the basis of this extension mechanism: https://marginaleffects.com/bonus/extensions.html

I bet most of the infrastructure is in place and it would be pretty easy. By default, we'd allow the types supplied by the upstream package, but when the model is "tagged" for compatbility (via s3 inheritance), we could trigger the prediction fuction supplied by the other package.

**Comment by garrettvandekamp** [comment]
Fun! I'll definitely keep a lookout then.

One package that comes to mind is `topmodels`, which I see from other posts you are already familiar with. Being able to predict the types from their `procast()` function would be pretty cool. Of course, their package at present only supports a limited number of models - far fewer than the number of models `marginaleffects` currently supports.

**Comment by vincentarelbundock** [comment]
Oh yeah! We have a pull request that was working a year ago and will be easy to merge for `topmodel` support

https://github.com/vincentarelbundock/marginaleffects/pull/1159
